### PR TITLE
add randomness to nodeStatusReportFrequency for kubelet

### LINKED ISSUE #124202: kubelet periodic node status update should not be aligned
### What happened?

The default `nodeStatusUpdateFrequency` is 10s with 4% jitter.
The default `nodeStatusReportFrequency` is 5m.

The node status will be updated either there's a change or it has been longer than `nodeStatusReportFrequency` since the previous update.
`nodeStatusReportFrequency` 
https://github.com/kubernetes/kubernetes/blob/55b83c92b3b69cd53d5bf22b8ccff859a005241a/pkg/kubelet/kubelet_node_status.go#L580-L581

The jitter of `nodeStatusUpdateFrequency` won't effectively make the periodic node status update to be spread evenly over the 5-minute window.
The node status update requests from thousands of nodes are still sent at almost the same time after a couple of days. It can cause the CPU usage to spike every 5 minutes.

The following is the APIServer CPU utilization on 96CPU instances. The cluster has 5k nodes that are added in a short period of time. The cluster is idle without any application workload. The utilization can go up to 64% and go down to 14% just because of these  node status update requests.
Screenshot 2024-04-05 at 12 59 16 PM

### What did you expect to happen?

I'd expect the node status update requests from kubelet to be spread evenly so that the apiserver won't have a minute that it has to handle the requests from most nodes every 5 minutes. 

### How can we reproduce it (as minimally and precisely as possible)?

Bring up a cluster and add nodes in a short period of time.

### Anything else we need to know?
Ref: https://github.com/kubernetes/kubernetes/pull/105272/files

_No response_

### Kubernetes version

<details>
Cutting edge 1.30
```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>
EKS
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by mengqiy**
Adding one time [-50% , +50%] randomness to `nodeStatusReportFrequency` after initial node status update.
It helps spread the load from kubelet evenly

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

The node status update traffic from kubelet can be almost synchronized in some scenarios and caused high CPU spikes. e.g. #124202

#### Which issue(s) this PR fixes:

Fixes #124202

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

Original release note (removed since this was rolled back
```
Add a one-time random duration of up to 50% of kubelet's nodeStatusReportFrequency to help spread the node status update load evenly over time.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by SergeyKanzhelev** [comment]
> Adding [-25% , +25%] randomness to `nodeStatusReportFrequency`.
> It helps spread the load from kubelet evenly

Would it be better to just add some random delay in the beginning once node is registered successfully the first time, instead of changing the reporting interval?

**Comment by SergeyKanzhelev** [comment]
> /milestone v1.32

@dims the bug is not blocking for release, it is not a regression, correct? Just checking why we are blocking release on it.

**Comment by dims** [comment]
/milestone clear

@SergeyKanzhelev was reflecting the request from the person working on the issue - https://github.com/kubernetes/kubernetes/issues/124202#issuecomment-2442663774

**Comment by mengqiy** [comment]
>Would it be better to just add some random delay in the beginning once node is registered successfully the first time, instead of changing the reporting interval?

A cross-cutting failure can possibly cause all nodes to become NotReady. When the failure is resolved, all kubelet changed status to be Ready at the same time and then all future node status update will become mostly synchronized again.  e.g. a VPC service outage or a user misconfiguration may cause CNI to make the node NotReady. When the issue is fixed, kubelet status report time are synchronized again which is undesired.

The idea is similar to how we add jitter to `nodeStatusUpdateFrequency`.
https://github.com/kubernetes/kubernetes/blob/66b3dc1a389d24620fc92ddfaca5f35190ef660d/pkg/kubelet/kubelet.go#L1684

**Comment by aojea** [comment]
we had several problems recently adjusting node reporting timeouts and some controllers setting the node unready or with some side effects so this should be carefully reviewed

**Comment by mengqiy** [comment]
@aojea Can you please link the GH issues that you mentioned above?

**Comment by SergeyKanzhelev** [comment]
I also vaguely remember something along these lines. Maybe https://github.com/kubernetes/kubernetes/issues/127352 triggered me.

But the point is - there may be some expectations on how fast node will reconcile status, so changing it even 25% may affect things negatively.

Adding a small jitter after `changed` triggered the update and keep subsequent updates in standard configured intervals may be a better approach to keep the predictability of node status update behavior.

**Comment by mengqiy** [comment]
Thanks for the pointer to the issue! This issue seems to be about nodeStatus**Update**Frequency but not nodeStatus**Report**Frequency.

From my understanding, there are 2 things in kubelet:
- nodeStatus**Update**Frequency: frequency to update lease object for the node. It has 10-second by default. If the lease not updated within `node-monitor-grace-period`, it will be set as NotReady by the node controller in kube-controller-manager.
- nodeStatus**Report**Frequency: frequency to update node/status subresource. It has 5-minute by default when there's no status change. Please note that kubelet will **immediate** update the node/status when there's a status change.

Given the kubelet **won't** wait for `nodeStatusReportFrequency` when there's a status change. IMO it should not be a problem if we make it to be 75% - 125% of the specified `nodeStatusReportFrequency`.

If you think this is still a strong concern, I'm happy to change it to be 75% - 100% of the specified `nodeStatusReportFrequency` or something like 80% - 100%.

Please let me know WDYT?

**Comment by SergeyKanzhelev** [comment]
Yes, I realize it is not the issue directly affecting this one. Have you looked into any logic k8s owns that may depend on this interval?

Do you think the proposal of a single jitter after a forced update works? Or you want to randomize the interval?

**Comment by mengqiy** [comment]
>Yes, I realize it is not the issue directly affecting this one. Have you looked into any logic k8s owns that may depend on this interval?

I haven't found anything depends on `nodeStatusReportFrequency `. I have tested with 30m with [-25%, 25%] jitter in a larger than 5k-node cluster setup. It seems to be working fine.

>Do you think the proposal of a single jitter after a forced update works? Or you want to randomize the interval?

To prevents kubelet status update from becoming synchronized again due to some crosscutting event as I described in https://github.com/kubernetes/kubernetes/pull/128394#issuecomment-2443363423, IMO it's more robust to always randomize the interval.

File: pkg/kubelet/kubelet.go
Comment (aojea) [inline pkg/kubelet/kubelet.go:1213]: I see, so `nodeStatusUpdateFrequency` is the one that has the dependencies on controllers and `nodeStatusReportFrequency` depends only on the leases ...

Comment (aojea) [reply pkg/kubelet/kubelet.go:1213]: why adding a new field here and not calculating it directly on `tryUpdateNodeStatus` loop?

File: pkg/kubelet/kubelet_node_status.go
Comment (aojea) [inline pkg/kubelet/kubelet_node_status.go:591]: if you initialize the value here the first time it will be zero, no? so `shouldPatchNodeStatus` will always trigger

**Comment by aojea** [comment]
> To prevents kubelet status update from becoming synchronized again due to some crosscutting event as I described in [#128394 (comment)](https://github.com/kubernetes/kubernetes/pull/128394#issuecomment-2443363423), IMO it's more robust to always randomize the interval.

I don't understand what do you mean by more robust, if any node starts in a random interval and the frequency is fixed , the randoms on the origin is maintained,  the other is completely random and more unpredictable for debugging, having deterministic patterns always help to troubleshoot problems because the outliers help to identify the issues

**Comment by mengqiy** [comment]
The relationship between them is:
 Kubelet uses 0.25 * `NodeLeaseDurationSeconds` as the renew interval. Currently the renew interval is 10s.
https://github.com/kubernetes/kubernetes/blob/6cb5ea56cb96f6a4ddec677d3d94faf07179368d/staging/src/k8s.io/kubelet/config/v1beta1/types.go#L284-L293

The node status will be set to `NotReady` if kubelet failed to renew its lease after `nodeMonitorGracePeriod`
https://github.com/kubernetes/kubernetes/blob/6cb5ea56cb96f6a4ddec677d3d94faf07179368d/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L270-L291

The `NodeStatusUpdateFrequency` is how frequently kubelet compute the node status. Kubelet only updates node status in APIServer in one of the following condition 1) there's a status change 2) lastStatusUpdateTime + `NodeStatusReportFrequency` < currentTime. 

https://github.com/kubernetes/kubernetes/blob/6cb5ea56cb96f6a4ddec677d3d94faf07179368d/pkg/kubelet/apis/config/types.go#L188-L192

https://github.com/kubernetes/kubernetes/blob/6cb5ea56cb96f6a4ddec677d3d94faf07179368d/pkg/kubelet/apis/config/types.go#L193-L197

Given 1) `nodeMonitorGracePeriod` doesn't depend on `NodeStatusReportFrequency` 2) when there's a node status change, the worst latency for kubelet to reflect it in nodeStatus is `NodeStatusUpdateFrequency` but not `NodeStatusReportFrequency` 3) this PR doesn't touch `NodeStatusUpdateFrequency`, the risk of this PR should low.

File: pkg/kubelet/kubelet_node_status.go
Comment (mengqiy) [reply pkg/kubelet/kubelet_node_status.go:591]: `kl.nextNodeStatusReportInterval` is first initialized in `pkg/kubelet/kubelet.go`. See:
https://github.com/kubernetes/kubernetes/pull/128394/files#diff-67dd9e8f3ee257072765326cb4f242852554a2c0753563fa51e292c0a63a7b94R562

File: pkg/kubelet/kubelet.go
Comment (mengqiy) [reply pkg/kubelet/kubelet.go:1213]: >why adding a new field here and not calculating it directly on tryUpdateNodeStatus loop?

Trying to ensure `nodeStatusReportFrequency` is bounded. IOW it is guaranteed that it won't be more than 125% of user specified `nodeStatusReportFrequency` value.

If we do `new nodeStatusReportFrequency = [75%, 125%] * old nodeStatusReportFrequency`, the worst case is that it keeps getting multiplied by 1.25 repeatedly and after long enough time, `nodeStatusReportFrequency` can become much significantly longer than the user specified `nodeStatusReportFrequency`. I'm trying to prevent this worst case scenario.

File: pkg/kubelet/kubelet_node_status.go
Comment (aojea) [inline pkg/kubelet/kubelet_node_status.go:582]: LGTM, just one question, maybe for @SergeyKanzhelev or other kubelet code maintainers, If we just want to add a constant to the period `kl.nodeStatusReportFrequency`, why don't we just add it here with a sync.Once so we avoid having to plumb the variable through different layers? 
This will make this solution specific of this control loop, and avoid to expose to the rest of the codebase

File: pkg/kubelet/kubelet_node_status.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/kubelet_node_status.go:585]: why do we need `!changed` here?

File: pkg/kubelet/kubelet_node_status.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/kubelet_node_status.go:586]: with this logic we are not fixing a situation when many nodes come back to ready at the same time. I thought we want to add randomness after every time `changed` was `true`

File: pkg/kubelet/kubelet_node_status.go
Comment (mengqiy) [reply pkg/kubelet/kubelet_node_status.go]: I have re-organize the code and hopefully it's easier to reason about.
```
	shouldPatchNodeStatus := changed || passStatusReportInterval
	if changed {
		// generate a random delay
	}

	if passStatusReportInterval {
		// reset nodeStatusReportWaitDurationAfterInitialUpdate to 0 when we use this random interval once.
	}

```

File: pkg/kubelet/kubelet_node_status.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/kubelet_node_status.go:584]: code lgtm. This is a good improvement.

Can you please add tests? If full function test is hard, maybe make it into the utility method that can be unit tested.

File: pkg/kubelet/kubelet_node_status.go
Comment (mengqiy) [reply pkg/kubelet/kubelet_node_status.go:582]: kubelet starts 2 goroutines in https://github.com/kubernetes/kubernetes/blob/b240c51a725d6c43221709c9b7a04f937757af37/pkg/kubelet/kubelet.go#L1677-L1684.
Both can execute the function `tryUpdateNodeStatus `. `sync.Once` won't give what we want if there are 2 node status update in a short period of time.

File: pkg/kubelet/kubelet_node_status.go
Comment (mengqiy) [reply pkg/kubelet/kubelet_node_status.go:584]: Thanks for the reivew!
Let me see what I can do for adding tests.

**Comment by mengqiy** [comment]
Added a unit test. PTAL

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

@aojea this should be ok. Please let me know if you see serious issues with this.

File: pkg/kubelet/kubelet_node_status.go
Comment (aojea) [reply pkg/kubelet/kubelet_node_status.go:582]: I see, I understand now , we reset in "each" update , not only once, thanks

File: pkg/kubelet/kubelet_node_status.go
Comment (aojea) [inline pkg/kubelet/kubelet_node_status.go:597]: I'm always scared about this concatenation of operations and relaying on knowing the order,  can you group the operations ?
```suggestion
	passStatusReportInterval := !kl.lastStatusReportTime.IsZero() && 
	   ( kl.clock.Since(kl.lastStatusReportTime) >= (kl.nodeStatusReportFrequency+kl.delayAfterNodeStatusChange) )
```

File: pkg/kubelet/kubelet_node_status.go
Comment (aojea) [inline pkg/kubelet/kubelet_node_status.go:596]: it is a bit overloaded for my taste, but we were back and forth so I don't want to stop progress, vur I think it will be more clear if you describe with comments all the steps and encapsulate only the logic to "isUpdateStatusPeriodExperid() bool" and "calculateDelay() time.Duration", so both are testable

if I understand now correctly the logic this should be something like
new func
```go
func (kl *kubelet) isUpdateStatusPeriodExperid bool {
if kl.lastStatusReportTime.IsZero() {
     return false
}
if kl.clock.Since(kl.lastStatusReportTime) >= (kl.nodeStatusReportFrequency+kl.delayAfterNodeStatusChange) {
   return true
}
return false
}
```
```go
// no need to update the status yet
if !changed && !kl.isUpdateStatusPeriodExperid() {
	kl.markVolumesFromNode(node)
        return nil
}
// We need to update the node status, if this is caused by a node change we want to calculate a new
// random delay so we avoid all the nodes to reach the apiserver at the same time. If the update is not related
// to a node change, because we run over the period, we reset the random delay so the node keeps updating
// its status at the same cadence
if changed {
     kl.delayAfterNodeStatusChange = kl.calculateDelay()
 else if {
     kl.delayAfterNodeStatusChange = 0
} 
updatedNode, err := kl.patchNodeStatus(originalNode, node)
 if err == nil {
		kl.markVolumesFromNode(updatedNode)
}

```

File: pkg/kubelet/kubelet_node_status.go
Comment (aojea) [reply pkg/kubelet/kubelet_node_status.go:597]: commented here https://github.com/kubernetes/kubernetes/pull/128394/files#r1829180689 , I think we can make this more readable

**Comment by SergeyKanzhelev** [comment]
unit test failure is new and seems to be introduced by this PR

**Comment by SergeyKanzhelev** [comment]
/remove-approve

to prevent accidental merge

**Comment by bart0sh** [comment]
> unit test failure is new and seems to be introduced by this PR

I can see it in another PR: https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128576/pull-kubernetes-e2e-kind-ipv6/1853939835513868288

**Comment by aojea** [comment]
> unit test failure is new and seems to be introduced by this PR

can be related? I think something related to sleep hooks merged ... the code here does not seem to interact with the pod lifecycle, or does it?

**Comment by SergeyKanzhelev** [comment]
> > unit test failure is new and seems to be introduced by this PR
> 
> can be related? I think something related to sleep hooks merged ... the code here does not seem to interact with the pod lifecycle, or does it?

yes, my mistake

File: pkg/kubelet/kubelet_node_status.go
Comment (mengqiy) [reply pkg/kubelet/kubelet_node_status.go:596]: Thanks for your suggestion!
I agree this looks better. I will make the change shortly.

**Comment by mengqiy** [comment]
@aojea @SergeyKanzhelev Updated and PTAL

**Comment by aojea** [comment]
/lgtm
/hold cancel

Thank you very much 

@SergeyKanzhelev for final approval

**Comment by aojea** [comment]
> unit test failure is new and seems to be introduced by this PR

@SergeyKanzhelev on additional note, it seems something got merged that is impacting this test https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-e2e-kind-alpha-beta-features/1853869116201373696

**Comment by liggitt** [comment]
this made TestUpdateNodeStatusWithLease *very* flaky

```
15s: 68 runs so far, 41 failures (60.29%), 10 active
```

rollback in https://github.com/kubernetes/kubernetes/compare/master...liggitt:kubernetes:revert-spreadkubeletload?expand=1

**Pull request merged**
